be/src/vec/exec/join/process_hash_table_probe_impl.h

```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
+    _build_block_offsets.emplace_back(block_offset);
+    _build_block_rows.emplace_back(block_row);
+    current_offset++;
+}
+
+template <int JoinOpType>
+template <bool need_null_map_for_probe, bool ignore_null, typename HashTableType,
+          bool with_other_conjuncts, bool is_mark_join>
 Status ProcessHashTableProbe<JoinOpType>::do_process(HashTableType& hash_table_ctx,
                                                      ConstNullMapPtr null_map,
                                                      MutableBlock& mutable_block,
-                                                     Block* output_block, size_t probe_rows,
-                                                     bool is_mark_join) {
+                                                     Block* output_block, size_t probe_rows) {
     auto& probe_index = *_join_context->_probe_index;
-    auto& probe_raw_ptrs = *_join_context->_probe_columns;
 
-    _probe_indexs.resize(_batch_size);
-    if (_build_block_rows.size() < probe_rows * PROBE_SIDE_EXPLODE_RATE) {
-        _build_block_rows.resize(probe_rows * PROBE_SIDE_EXPLODE_RATE);
-        _build_block_offsets.resize(probe_rows * PROBE_SIDE_EXPLODE_RATE);
-    }
     using KeyGetter = typename HashTableType::State;
     using Mapped = typename HashTableType::Mapped;
 
-    int right_col_idx =
-            _join_context->_is_right_semi_anti ? 0 : _join_context->_left_table_data_types->size();
-    int right_col_len = _join_context->_right_table_data_types->size();
-
-    KeyGetter key_getter(probe_raw_ptrs, _join_context->_probe_key_sz, nullptr);
-
-    if constexpr (ColumnsHashing::IsPreSerializedKeysHashMethodTraits<KeyGetter>::value) {
-        if (probe_index == 0) {
-            _pre_serialize_key(probe_raw_ptrs, probe_rows, _probe_keys);
-        }
-        key_getter.set_serialized_keys(_probe_keys.data());
-    }
+    KeyGetter key_getter = _init_probe_side<KeyGetter>(probe_rows, with_other_conjuncts);
 
     auto& mcol = mutable_block.mutable_columns();
-    int current_offset = 0;
 
     constexpr auto is_right_semi_anti_join =
             JoinOpType == TJoinOp::RIGHT_ANTI_JOIN || JoinOpType == TJoinOp::RIGHT_SEMI_JOIN;
 
     constexpr auto probe_all =
             JoinOpType == TJoinOp::LEFT_OUTER_JOIN || JoinOpType == TJoinOp::FULL_OUTER_JOIN;
 
-    bool all_match_one = true;
     int last_probe_index = probe_index;
+
+    int current_offset = 0;
+    bool all_match_one = true;
     size_t probe_size = 0;
-    auto& probe_row_match_iter =
-            std::get<ForwardIterator<Mapped>>(*_join_context->_probe_row_match_iter);
+    auto& probe_row_match_iter = _probe_row_match<Mapped, with_other_conjuncts>(
+            current_offset, probe_index, probe_size, all_match_one);
+
+    bool is_the_last_sub_block = false;
+    if (with_other_conjuncts && probe_size != 0) {
+        is_the_last_sub_block = !probe_row_match_iter.ok();
+        _same_to_prev.emplace_back(false);
```

Comment:
if not other join conjunt do not need do the work

Reply:
This branch will be entered only when with_other_conjuncts is true